### PR TITLE
Improve support for the Alva BC6 braille displays

### PR DESCRIPTION
**Pull request by dkager**
Related issue: #5206

This adds a bunch of default key assignments for the Alva BC6 braille displays, approved by Optelec. Most commands were already in NVDA, but I added two new ones that are not merged into this branch. See the commit messages for the relevant issue numbers.

### CODE REVIEW

**Comment by gregjozk** [comment]
Hi,

does anyone know, when will this branch be included into main nvda branch?

i'm asking this because I have got a problem and hope that it can solve that one.

here is a description:

whenever i've tried to change braille display to alva bc6, i saw a message: Could not load Alva bc6 display.

so has main nvda got a bug (alvaw32.dll is older than one on main optelec's site) or do i make something wrong?

this problem was observed both on 32-bit system and 64-bit systems.

thanks, J
[nvda.zip](https://github.com/nvaccess/nvda/files/278646/nvda.zip)

see from line 219 to line 256.

**Comment by dkager** [comment]
Are you connecting via USB or Bluetooth?

The included DLL causes no problems for me, but it may be worth it to look at adding the latest version, assuming Optelec is distributing it.

**Comment by gregjozk** [comment]
Hi,

i'm connecting via usb, cause protocol converter does not support bt.

thx, J

2016-05-24 7:16 GMT+02.00, Davy Kager <notifications@github.com>:

**Comment by jcsteh** [comment]
We're very deliberately not using the latest version because it has a 
compulsory dependency on a whole load of extra libraries. I can't 
remember the exact size, but I seem to recall that distributing the bare 
minimum would be over 10 mb, which is totally unacceptable for a single 
display. If updating to the latest version to fix a bug proved to be 
essential, I think we'd need to drop support for this display from the 
core and bundle it as an add-on.

**Comment by gregjozk** [comment]
i thought, that this driver supports bc640/680, 640 comfort, 640 usb
comfort and also protocol converter. it seems, i was wrong.

2016-05-24 7:31 GMT+02.00, James Teh <notifications@github.com>:

**Comment by jcsteh** [comment]
It does support all of those, so yes, saying "one display" isn't valid. 
However, it's still only one range of displays, and IMO, that still 
doesn't justify growing NVDA by 10 mb.

**Comment by gregjozk** [comment]
i understand.

unfortunately, i cannot help you develop an addon because, i don't
have enough programming skils.

thanks.

2016-05-24 7:57 GMT+02.00, James Teh <notifications@github.com>:

**Comment by dkager** [comment]
Which exact display are you using? My BC640 never had problems when connecting over USB on Windows 7 or 10. But I gather you are using the protocol converter. Are Optelec themselves aware of any issues?

**Comment by gregjozk** [comment]
Hi,

as I and my local dealer know, at Optelec aren't aware of any issues
with protocol converter and nvda when using alva satellite (544 pro),
which I am using. yes, I know, that I can use brltty, but as I know it
is not supported on 64-bit systems.

one question: does problem exists, because a spacific service is not
loaded and is not running or what is wrong that this problem occurs?

during testing a converter on 32-bit windows 7, where a display is
usually used, i was playing with brlapi service, and i set it tu start
with delay, which caused a same problem as described earlier. reset
service to automatic start has solved an issue, but i don't know,
which service should be loaded for bc6 series.

on 64-bit windows 10 machine brltty is not installed.

thanks, Jožef

2016-05-24 15:30 GMT+02.00, Davy Kager <notifications@github.com>:

**Comment by jcsteh** [comment]
No; we don't use any service to communicate with ALVA displays.

File: user_docs/en/userGuide.t2t [hunk 1]
```diff
@@ -1570,11 +1570,17 @@ Following are the key assignments for this display with NVDA.
 Please see the display's documentation for descriptions of where these keys can be found.
 %kc:beginInclude
 || Name | Key |
-| Scroll braille display back | t1 |
+| Scroll braille display back | t1 or etouch1 |
 | Move braille display to previous line | t2 |
+| Move to focus object | t3 |
```
Comment (jcsteh) [inline user_docs/en/userGuide.t2t]: We don't mention "object" elsewhere for this command. Also, it moves to the cursor, not just to the object. I suggest this be changed to "Move to current focus".

File: user_docs/en/userGuide.t2t [hunk 2]
```diff
@@ -1570,11 +1570,17 @@ Following are the key assignments for this display with NVDA.
...
+| Scroll braille display back | t1 or etouch1 |
 | Move braille display to previous line | t2 |
+| Move to focus object | t3 |
 | Move braille display to next line | t4 |
-| Scroll braille display forward | t5 |
+| Scroll braille display forward | t5 or etouch3 |
 | Route to braille cell | routing |
+| Move to top line in review | t1+t2 |
+| Move to bottom line in review | t4+t5 |
+| Braille Tethered to | t1+t3 |
```
Comment (jcsteh) [inline user_docs/en/userGuide.t2t]: For consistency, this should be "Toggle braille tethered to"

**Comment by jcsteh** [comment]
Merge note: Happy for this to be incubated and then merged to master once trivial User Guide changes are made; see review comments.

What's New entry: in New Features:

```
- Additional key assignments have been added for ALVA BC640/680 series braille displays. (#5206)
- NVDA now correctly detects the secondary routing buttons on ALVA BC640/680 series braille displays. (#5206)
```

**Comment by dkager** [comment]
Applied both user docs tweaks.

**Comment by michaelDCurran** [comment]
Can a what's new entry please be provided on this pr as a comment, then I can merge it to master.

**Comment by michaelDCurran** [comment]
Never mind. Saw it.

**Pull request merged**